### PR TITLE
#0: Make CoreRangeSet accept/store CoreRanges as a vector internally to support user specified ordering

### PR DESCRIPTION
**Pull request by tt-aho**
### Ticket
Link to Github Issue

### Problem description
We need to support user specified ordering of cores in a CoreRangeSet to support Llama's use of SubCoreMesh and how the cores are sorted. Currently we store the core ranges internally as a set, which will sort the cores top left to bottom right, so we want to swap to an insertion order container.
Additional change is to split core_coord.h into separate implementation and header files.

### What's changed
Change the internal storage of the CoreRangeSet from set to vector. Add new constructors to take in vector, single CoreRange, and default constructor since these are common use cases throughout the codebase.
Split core_coord.h into separate implementation and header files.

### Checklist
- [x] Post commit CI passes
Post Commit: https://github.com/tenstorrent/tt-metal/actions/runs/11462061298
T3K Freq: https://github.com/tenstorrent/tt-metal/actions/runs/11465611957
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
Single Device: https://github.com/tenstorrent/tt-metal/actions/runs/11462066763
T3K: https://github.com/tenstorrent/tt-metal/actions/runs/11462077357
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by bbradelTT** [comment]
Since this is not in the description:
1. Would we want to change the name of the class to e.g. CoreRanges to reflect that the class is no longer a set?
2. Are there any performance concerns with changing the time complexity of some operations?

**Comment by tt-aho** [comment]
> Since this is not in the description:
> 
> 1. Would we want to change the name of the class to e.g. CoreRanges to reflect that the class is no longer a set?
> 2. Are there any performance concerns with changing the time complexity of some operations?

1. I'd still consider it a set in a sense that there are no duplicate cores. We assert out if any ranges overlap which is why I left the name as CoreRangeSet.

2. Iteration over the ranges of a CoreRangeSet should have improved performance. The main two performance concerns are searching and if we depend on sorting. From my initial changes, there was only one usage of search I saw in `ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_dram_sharded_program_factory.cpp`, which is during program creation code which I don't think we consider that performance critical. The main dependency on sorting is for the CoreRangeSet merge function, which I believe is also mainly used in the infrastructure for setting up a program and isn't on the perf critical path. Inside merge we copy the internal container into a new set, (this was also the old behaviour when we internally stored as set), so I don't think there was a complexity change here.

**Review by bbradelTT — APPROVED** [review]

**Review by yugaoTT — APPROVED** [review]
looks great! Thanks @tt-aho

**Review by ntarafdar — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]

**Review by abhullar-tt — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

File: tt_metal/common/core_coord.hpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+    friend void swap(CoreRangeSet &first, CoreRangeSet &second);
+
+    CoreRangeSet(const CoreRangeSet &other);
+
+    CoreRangeSet &operator=(const CoreRangeSet &other);
+
+    CoreRangeSet(CoreRangeSet &&other);
+
+    CoreRangeSet &operator=(CoreRangeSet &&other);
```
Comment (patrickroberts) [inline tt_metal/common/core_coord.hpp:132]: Can you add a move constructor from vector?
```cpp
CoreRangeSet(std::vector<CoreRange> &&core_ranges) noexcept;

...

CoreRangeSet::CoreRangeSet(std::vector<CoreRange> &&core_ranges) noexcept : ranges_(std::move(core_ranges)) {}
```

Comment (tt-aho) [reply tt_metal/common/core_coord.hpp:132]: Done

**Review by patrickroberts — APPROVED** [review]
At some point I was hoping to implement short buffer optimization (SBO) on `CoreRangeSet` to avoid allocation at least in the cases when storing a single `CoreCoord` or `CoreRange`. I think that can still eventually be done on top of these changes, which is the only thing I was concerned about when reviewing this. Looks good, and thanks for the improvement.

**Review by SeanNijjar — APPROVED** [review]
Only reviewed CCL portion

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**